### PR TITLE
Detect compute shaders with only indirect calls

### PR DESCRIPTION
**Pull request by Flakebi**
For an indirect call, no function is called so iterating over function
definitions only would not detect the call and lead to the wrong
pipeline layout.

Only the topmost commit “Detect compute shaders with only indirect calls” is part of this pr.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit e5a4383d7ec24529c4058d8ff08a2a42c92d1ae3
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev eb28c7ef53329f838e693f233696823330e964c0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev fde3d4e2bd90743606ca7b724bf62a53cacf4615</li>
</ul></details>

<details><summary>CTS tests (Failed: 2/144330)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26320/48174 (54.6%)</li>
<li><details><summary>Failed: 1/48174 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.spirv_assembly.instruction.graphics.float_controls.fp64.input_args.length_denorm_flush_to_zero_vert
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 21853/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24852/48153 (51.6%)</li>
<li>Failed: 0/48153 (0.0%)</li>
<li>Skipped: 23301/48153 (48.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx103</summary><ul>

<li>Passed: 26218/48003 (54.6%)</li>
<li><details><summary>Failed: 1/48003 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.spirv_assembly.instruction.graphics.float_controls.fp64.input_args.length_denorm_flush_to_zero_vert
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 21784/48003 (45.4%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchEntryPointMutate.cpp [hunk 1]
```diff
@@ -291,7 +291,26 @@ void PatchEntryPointMutate::setupComputeWithCalls(Module *module) {
     if (func.isDeclaration() && func.getIntrinsicID() == Intrinsic::not_intrinsic &&
         !func.getName().startswith(lgcName::InternalCallPrefix) && !func.user_empty()) {
       m_computeWithCalls = true;
-      break;
+      return;
+    }
+
+    // Search for indirect calls
+    if (!func.isDeclaration()) {
```
Comment (trenouf) [inline lgc/patch/PatchEntryPointMutate.cpp:298]: You don't need this "if"; for a declaration, the following loop has zero iterations.

File: lgc/patch/PatchEntryPointMutate.cpp [hunk 2]
```diff
@@ -291,7 +291,26 @@ void PatchEntryPointMutate::setupComputeWithCalls(Module *module) {
...
+      for (const BasicBlock &block : func) {
+        for (const Instruction &inst : block) {
+          if (auto *call = dyn_cast<CallInst>(&inst)) {
+            Value *calledVal = call->getCalledOperand();
+            if (auto calledFunc = dyn_cast<Function>(calledVal)) {
```
Comment (trenouf) [inline lgc/patch/PatchEntryPointMutate.cpp:303]: You only need to detect indirect calls here, so how about just checking !isa<Function>(calledVal)? Then you don't need the next few lines.

**Review by trenouf — APPROVED** [review]
LGTM modulo minor niggles.

File: lgc/patch/PatchEntryPointMutate.cpp (hunk 2 above)
Comment (Flakebi) [reply lgc/patch/PatchEntryPointMutate.cpp]: Thanks, that looks better

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit a101d7787d6f7276f92b4c69edde1d0383d6ffb3
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev eb28c7ef53329f838e693f233696823330e964c0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 529059d0afd8e37f444592b25b3fb6a3d628ef89</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/0)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit a101d7787d6f7276f92b4c69edde1d0383d6ffb3
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev eb28c7ef53329f838e693f233696823330e964c0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 529059d0afd8e37f444592b25b3fb6a3d628ef89</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/0)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@Flakebi please resolve conflicts

**Comment by Flakebi** [comment]
This is also part of #1071, so I’ll close this one.

**Pull request closed without merging**